core/interop/src/lib.rs

```diff
@@ -40,67 +46,251 @@ impl<T: IntoIterator<Item = (JsString, NativeFunction)> + Clone> IntoJsModule fo
 /// This trait does not require the implementing type to be `Copy`, which
 /// can lead to undefined behaviour if it contains Garbage Collected objects.
 ///
-/// # Safety
-/// For this trait to be implemented safely, the implementing type must not contain any
-/// garbage collected objects (from [`boa_gc`]).
-pub unsafe trait IntoJsFunctionUnsafe {
+/// This trait is implemented for functions with various signatures.
+///
+/// For example:
+/// ```
+/// # use boa_engine::{Context, JsValue, NativeFunction};
+/// # use boa_interop::UnsafeIntoJsFunction;
+/// # let mut context = Context::default();
+/// let f = |a: i32, b: i32| a + b;
+/// let f = unsafe { f.into_js_function_unsafe(&mut context) };
+/// let result = f.call(&JsValue::undefined(), &[JsValue::from(1), JsValue::from(2)], &mut context).unwrap();
+/// assert_eq!(result, JsValue::new(3));
+/// ```
+///
+/// Since the `IntoJsFunctionUnsafe` trait is implemented for `FnMut`, you can
+/// also use closures directly:
+/// ```
+/// # use boa_engine::{Context, JsValue, NativeFunction};
+/// # use boa_interop::UnsafeIntoJsFunction;
+/// # use std::cell::RefCell;
+/// # use std::rc::Rc;
+/// # let mut context = Context::default();
+/// let mut x = Rc::new(RefCell::new(0));
+/// // Because NativeFunction takes ownership of the closure,
+/// // the compiler cannot be certain it won't outlive `x`, so
+/// // we need to create a `Rc<RefCell>` and share it.
+/// let f = unsafe {
+///     let x = x.clone();
+///     move |a: i32| *x.borrow_mut() += a
+/// };
+/// let f = unsafe { f.into_js_function_unsafe(&mut context) };
+/// f.call(&JsValue::undefined(), &[JsValue::from(1)], &mut context).unwrap();
+/// f.call(&JsValue::undefined(), &[JsValue::from(4)], &mut context).unwrap();
+/// assert_eq!(*x.borrow(), 5);
+/// ```
+pub trait UnsafeIntoJsFunction<Args, Ret>: private::IntoJsFunctionSealed<Args, Ret> {
     /// Converts the type into a JS function.
     ///
     /// # Safety
     /// This function is unsafe to ensure the callee knows the risks of using this trait.
     /// The implementing type must not contain any garbage collected objects.
-    unsafe fn into_js_function(self, context: &mut Context) -> NativeFunction;
+    unsafe fn into_js_function_unsafe(self, context: &mut Context) -> NativeFunction;
```

Comment:
I'll rename the copy form in `..._copied` but I think we should keep `_unsafe`. I've been using these helpers in my project and sometimes it does get confusing in the error messages (mostly because the type system just name the type with the file name and lines).

Reply:
Can you give some examples of that? I cannot see immediately why adding the `_unsafe` suffix would improve error messages.

Reply:
I don't really want to block the PR, so I'm resolving this. We can bikeshed later.